### PR TITLE
Optimize nearest neighb

### PR DESCRIPTION
**Pull request by Ghazi-Bouabene**
Added efficient nearest neighbour interpolation of discrete GP (efficient on the discrete points)

### CODE REVIEW

File: src/main/scala/scalismo/statisticalmodel/DiscreteLowRankGaussianProcess.scala [hunk 1]
```diff
@@ -220,7 +220,7 @@ case class DiscreteLowRankGaussianProcess[D <: Dim: NDSpace, DO <: Dim: NDSpace]
    * Interpolates discrete Gaussian process to have a new, continuous representation as a [[DiscreteLowRankGaussianProcess]],
    * using nearest neigbor interpolation (for both mean and covariance function)
    */
-  override def interpolateNearestNeighbor: LowRankGaussianProcess[D, DO] = {
+  override def interpolateNearestNeighbor: NearestNeighbourInterpolatedLowRankGaussianProcess[D, DO] = {
```
Comment (marcelluethi) [inline src/main/scala/scalismo/statisticalmodel/DiscreteLowRankGaussianProcess.scala]: This should not return a NearestNeighborInterpolatedLowRankGaussianProcess, but a LowRankGaussianProcess, as the former is a private class and should not be exposed in the public API (I am surprised that you don't get a warning from scala).

File: src/main/scala/scalismo/statisticalmodel/LowRankGaussianProcess.scala [hunk 2]
```diff
@@ -164,6 +164,32 @@ class LowRankGaussianProcess[D <: Dim: NDSpace, DO <: Dim: NDSpace](mean: Vector
...
+ */
+private[scalismo] class NearestNeighbourInterpolatedLowRankGaussianProcess[D <: Dim: NDSpace, DO <: Dim: NDSpace](mean: VectorField[D, DO], klBasis: KLBasis[D, DO], discreteGP: DiscreteLowRankGaussianProcess[D, DO]) extends LowRankGaussianProcess[D, DO](mean, klBasis) {
+
+  require(klBasis.size == discreteGP.rank)
+
+  override def instance(c: DenseVector[Float]): VectorField[D, DO] = discreteGP.instance(c).interpolateNearestNeighbor
+
+  // if all training data points belong to the interpolated discrete domain, then we compute a discrete posterior GP and interpolate it
+  // this way the posterior will also remain very efficient when sampling.
+  override def posterior(trainingData: IndexedSeq[(Point[D], Vector[DO], NDimensionalNormalDistribution[DO])]): LowRankGaussianProcess[D, DO] = {
```
Comment (marcelluethi) [inline src/main/scala/scalismo/statisticalmodel/LowRankGaussianProcess.scala]: cool that we can also make this one faster. Nice!

File: src/main/scala/scalismo/statisticalmodel/LowRankGaussianProcess.scala [hunk 3]
```diff
@@ -164,6 +164,32 @@ class LowRankGaussianProcess[D <: Dim: NDSpace, DO <: Dim: NDSpace](mean: Vector
 }
 
 /**
+ * Convenience class to speedup sampling from a LowRankGaussianProcess obtained by nearest neighbor interpolation of a DiscreteLowRankGaussianProcess
+ *
+ */
+private[scalismo] class NearestNeighbourInterpolatedLowRankGaussianProcess[D <: Dim: NDSpace, DO <: Dim: NDSpace](mean: VectorField[D, DO], klBasis: KLBasis[D, DO], discreteGP: DiscreteLowRankGaussianProcess[D, DO]) extends LowRankGaussianProcess[D, DO](mean, klBasis) {
```
Comment (marcelluethi) [inline src/main/scala/scalismo/statisticalmodel/LowRankGaussianProcess.scala]: I would move this code to the DiscreteLowRankGaussianProcess.scala file, as it is an implementation detail of that class. It should, in my opinion, nowhere else be visible.

**Comment by marcelluethi** [comment]
@Ghazi-Bouabene Very cool! This will speed up computations a lot. 
Besides the few minor comments, it seems good to merge.

**Comment by Ghazi-Bouabene** [comment]
Thanks. Just updated the PR accordingly.

**Pull request merged**